sensor_msgs/PointCloud2 space
geometry_msgs/Point[] codebook
